group/assignment5/main.cpp

```diff
@@ -0,0 +1,194 @@
+    vector<char> alphabet;
+    bool result;
+    unordered_map<char, int> idxs;
+    vector<bool> expectedChars;
+    vector<bool> actualChars;
+
+    // test 1 - assignment example
+    idxs.clear();
+    expectedChars = vector<bool>(pow(2, CHAR_BIT), false);
+    actualChars = vector<bool>(pow(2, CHAR_BIT), false);
+    result = findAlphabet(dictionary, alphabet);
+    EXPECT_TRUE(result);
+    EXPECT_EQ(4, alphabet.size());
+    for (char c : {'A', 'R', 'C', 'T'}) {
+        expectedChars[c] = true;
+    }
+    for (int i = 0; i < alphabet.size(); i++) {
+        idxs[alphabet[i]] = i;
+        actualChars[alphabet[i]] = true;
+    }
+    EXPECT_EQ(expectedChars, actualChars);
+    EXPECT_TRUE(idxs['A'] < idxs['R']);
+    EXPECT_TRUE(idxs['R'] < idxs['C']);
+    EXPECT_TRUE(idxs['T'] < idxs['R']);
+
+    // test 2 - empty dictionary
+    dictionary.clear();
+    result = findAlphabet(dictionary, alphabet);
+    EXPECT_TRUE(result);
+    EXPECT_EQ(0, alphabet.size());
+
+    // test 3 - inconsistent dictionary
+    dictionary = {"ART", "RAT", "ABS"};
+    result = findAlphabet(dictionary, alphabet);
+    EXPECT_FALSE(result);
+    EXPECT_EQ(0, alphabet.size());
+
+    // test 4
+    dictionary = {"ABC", "DEFG", "HIJK", "LMN", "OPQRS", "TU", "VWXYZ"};
+    expectedChars = vector<bool>(pow(2, CHAR_BIT), false);
+    actualChars = vector<bool>(pow(2, CHAR_BIT), false);
+    idxs.clear();
+    result = findAlphabet(dictionary, alphabet);
+    EXPECT_TRUE(result);
+    EXPECT_EQ(26, alphabet.size());
+    for (char c = 'A'; c <= 'Z'; c++) {
+        expectedChars[c] = true;
+    }
+    for (int i = 0; i < alphabet.size(); i++) {
+        idxs[alphabet[i]] = i;
+        actualChars[alphabet[i]] = true;
+    }
+    EXPECT_EQ(expectedChars, actualChars);
+    EXPECT_TRUE(idxs['A'] < idxs['D']);
+    EXPECT_TRUE(idxs['D'] < idxs['H']);
+    EXPECT_TRUE(idxs['H'] < idxs['L']);
+    EXPECT_TRUE(idxs['L'] < idxs['O']);
+    EXPECT_TRUE(idxs['O'] < idxs['T']);
+    EXPECT_TRUE(idxs['T'] < idxs['V']);
+
+    // test 5
+    dictionary = {"abd", "abc", "^pap", "^alk", "*$llo", "car", "01o"};
+    expectedChars = vector<bool>(pow(2, CHAR_BIT), false);
+    actualChars = vector<bool>(pow(2, CHAR_BIT), false);
+    idxs.clear();
+    result = findAlphabet(dictionary, alphabet);
+    EXPECT_TRUE(result);
+    EXPECT_EQ(14, alphabet.size());
+    for (char c : {'a', 'b', 'c', 'd', 'p', '^', 'l', 'k', '*', '$', 'o', 'r', '0', '1'}) {
+        expectedChars[c] = true;
+    }
+    for (int i = 0; i < alphabet.size(); i++) {
+        idxs[alphabet[i]] = i;
+        actualChars[alphabet[i]] = true;
+    }
+    EXPECT_EQ(expectedChars, actualChars);
+    EXPECT_TRUE(idxs['a'] < idxs['^']);
+    EXPECT_TRUE(idxs['d'] < idxs['c']);
+    EXPECT_TRUE(idxs['p'] < idxs['a']);
+    EXPECT_TRUE(idxs['^'] < idxs['*']);
+    EXPECT_TRUE(idxs['*'] < idxs['c']);
+    EXPECT_TRUE(idxs['c'] < idxs['0']);
+
+    return 0;
+}
```

Comment:
Great solution! I like tests, they are very extensive. I can suggest only to encapsulate the logic inside of the class and provide only necessary interface for users. :)